readme.md
<p align="center">
  A simple animated menu for React Native apps.
</p>

# react-native-side-nav

![platforms](https://img.shields.io/badge/platforms-iOS-brightgreen.svg?style=flat-square&colorB=b91d73)
[![Package version](https://img.shields.io/npm/v/react-native-side-nav.svg?style=flat-square&colorB=b91d73)](https://npmjs.org/package/react-native-side-nav)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square&colorB=b91d73)](http://makeapullrequest.com)
[![npm downloads](https://img.shields.io/npm/dm/react-native-side-nav.svg?style=flat-square&colorB=b91d73)](https://npmjs.org/package/react-native-side-nav)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg?style=flat-square&colorB=b91d73)](https://opensource.org/licenses/MIT)
[![TypeScript](https://img.shields.io/badge/%3C%2F%3E-TypeScript-%230074c1.svg)](http://www.typescriptlang.org/)

## Contents

- [Contents](#contents)
- [Usage](#usage)
- [Props](#props)
- [Examples](#examples)
- [Roadmap](#roadmap)

## Usage

```jsx
import React from "react";
import { View, TouchableOpacity } from "react-native";
import SideMenu from "react-native-side-nav";

const MenuComponent = () => {
  return <View style={styles.animatedBox}>
         // Your menu content here
         </View>;
};

const App = () => {
  const [menuExpanded, setMenuExpanded] = useState(false);

  return (
    <View>
      <SideMenu
        menuExpanded={menuExpanded}
        menuComponent={<MenuComponent />}
        burgerIcon
        onToggle={() => setMenuExpanded(!menuExpanded)}
      >
        // Your App components
      </SideMenu>
    </View>
  );
};

export default App;
```

## Props

| Prop                                     |  Default  |    Type     | Description                                                                                                                                                                                         |
| :--------------------------------------- | :-------: | :---------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| menuExpanded                             |   false   |  `boolean`  | Set to `true` or `false` to open or close the menu.                                                                                                                                                 |
| menuComponent                            |           | `ReactNode` | The content of your drawer. (see example)                                                                                                                                                           |
| menuWidth _optional_                     |    80     |  `number`   | The width of the menu, as a percentage of the screen width.                                                                                                                                         |
| animationDuration _optional_             |    300    |  `number`   | The length of the animation in ms.                                                                                                                                                                  |
| fade _optional_                          |   false   |  `boolean`  | Set to `true` to make the menu fade in on top of the content. Only works if `push` is set to `false`.                                                                                               |
| overlay _optional_                       |   false   |  `boolean`  | Set to `true` adds a dark overlay on top of the app when the menu is open.                                                                                                                          |
| overlayOpacity _optional_                |    0.4    |  `number`   | Sets the opacity of the overlay. Only works if `overlay` is set to `true`.                                                                                                                          |
| leftAligned _optional_                   |   false   |  `boolean`  | Set to `true` to make the menu come in from the left side of the screen.                                                                                                                            |
| push _optional_                          |   false   |  `boolean`  | Set to `true` to make the menu 'push' your app contents to one side.                                                                                                                                |
| burgerIcon _optional_                    |   false   |  `boolean`  | Set to `true` to add an absolutely positioned burger icon to toggle your menu.                                                                                                                      |
| onToggle _required if burgerIcon={true}_ |   null    | `() => any` | Your toggle function for the burger icon. (see example)                                                                                                                         |
| onHide _optional_                        |   null    | `() => any` | The menu's hide function, pass this down if you want users to be able to close the menu by tapping outside of it. Ex `onHide={() => setShow(false)`                                                                                                                         |
| burgerWidth _optional_                   |    35     |  `number`   | The width of the burger icon.                                                                                                                                                                       |
| burgerIconStyles _optional_              |    {}     | `ViewStyle` | Edit the burger X and Y position. Example: `burgerIconStyles={{top: 30, left: 20}}`. You can try passing other styles although they haven't been tested. Use the burger icon color props instead.   |
| burgerIconColor1 _optional_              | '#ffffff' |  `string`   | The color you want your burger icon to be.                                                                                                                                                          |
| burgerIconColor2 _optional_              |    ''     |  `string`   | If the background of your menu doesn't match the rest of your app, you can change the color of the icon when the menu is expanded. If you don't set this value it will default to burgerIconColor1. |

## Examples

#### menuWidth

<div >
	<img src="img/menuWidth.gif" alt="" width="200px">
</div>

#### fade

<div >
	<img src="img/fade.gif" alt="" width="200px">
</div>

#### overlay & overlayOpacity

<div >
	<img src="img/overlay.gif" alt="" width="200px">
</div>

#### leftAligned

<div >
	<img src="img/leftAligned.gif" alt="" width="200px">
</div>

#### push

<div >
	<img src="img/push.gif" alt="" width="200px">
</div>

#### burgerIcon

<div >
	<img src="img/burgerIcon.gif" alt="" width="200px">
</div>

#### burgerWidth

<div >
	<img src="img/burgerWidth.gif" alt="" width="200px">
</div>

#### burgerIconStyles

<div >
	<img src="img/burgerIconStyles.gif" alt="" width="200px">
</div>

#### burgerIconColor1

<div >
	<img src="img/burgerIconColor1.gif" alt="" width="200px">
</div>

#### burgerIconColor2

<div >
	<img src="img/burgerIconColor2.gif" alt="" width="200px">
</div>

## Roadmap

- [ ] Improve animation
- [ ] Android support
- [ ] Update docs examples
- [x] Typescript
- [x] Add burger icon
- [x] Testing
- [x] Add overlay opacity prop

## Questions

Feel free to [contact me](mailto:serena.antonetti@gmail.com) or [create an issue](https://github.com/serenastorm/react-native-side-nav/issues/new)

> Made with ♥.


CHANGELOG.md
# Changelogs

## 2.0.3

  - Migrated to Typescript!
  - Updated the usage section of the docs
  - Added `onHide` prop to close drawer when tapping outside

## 2.0.0
  - Added burgerIcon

SideMenu.js
import React from "react";
import { Animated, Dimensions, StyleSheet, View } from "react-native";
import PropTypes from "prop-types";
import BurgerIcon from "./BurgerIcon";
import { TouchableOpacity } from "react-native-gesture-handler";

const screen = Dimensions.get("window");

class SideMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: false,
      fadeAnim: new Animated.Value(0.3),
      overlayAnim: new Animated.Value(0),
      leftOffset: new Animated.Value(0),
      rightOffset: new Animated.Value(screen.width),
      rightOffsetPush: new Animated.Value(0)
    };
  }

  openDrawer = () => {
    const { menuWidth, animationDuration, overlayOpacity } = this.props;
    const {
      leftOffset,
      rightOffset,
      rightOffsetPush,
      fadeAnim,
      overlayAnim
    } = this.state;
    const DRAWER_WIDTH = screen.width * (menuWidth / 100);

    Animated.parallel([
      Animated.timing(leftOffset, {
        toValue: DRAWER_WIDTH,
        duration: animationDuration
      }),
      Animated.timing(rightOffset, {
        toValue: -DRAWER_WIDTH,
        duration: animationDuration
      }),
      Animated.timing(rightOffsetPush, {
        toValue: -DRAWER_WIDTH,
        duration: animationDuration
      }),
      Animated.timing(fadeAnim, {
        toValue: 1,
        duration: animationDuration
      }),
      Animated.timing(overlayAnim, {
        toValue: overlayOpacity,
        duration: animationDuration
      })
    ]).start();
  };

  closeDrawer = () => {
    const { animationDuration } = this.props;
    const {
      leftOffset,
      rightOffset,
      rightOffsetPush,
      fadeAnim,
      overlayAnim
    } = this.state;

    Animated.parallel([
      Animated.timing(leftOffset, {
        toValue: 0,
        duration: animationDuration
      }),
      Animated.timing(rightOffset, {
        toValue: screen.width,
        duration: animationDuration
      }),
      Animated.timing(rightOffsetPush, {
        toValue: 0,
        duration: animationDuration
      }),
      Animated.timing(fadeAnim, {
        toValue: 0,
        duration: animationDuration
      }),
      Animated.timing(overlayAnim, {
        toValue: 0,
        duration: animationDuration
      })
    ]).start();
  };

  componentDidUpdate() {
    const { menuExpanded, burgerIconColor1, burgerIconColor2 } = this.props;

    menuExpanded ? this.openDrawer() : this.closeDrawer();
  }

  renderOverlay = () => {
    const {
      children,
      menuComponent,
      menuWidth,
      menuExpanded,
      overlay,
      leftAligned,
      burgerIcon,
      burgerIconColor1,
      burgerIconColor2,
      animationDuration,
      burgerIconStyles,
      burgerWidth
    } = this.props;
    const { fadeAnim, overlayAnim, rightOffset, leftOffset } = this.state;
    const drawerAnimation = () => {
      if (leftAligned) {
        return {
          transform: [{ translateX: leftOffset }],
          left: -DRAWER_WIDTH
        };
      } else {
        return {
          transform: [{ translateX: rightOffset }],
          left: screen.width
        };
      }
    };
    const DRAWER_WIDTH = screen.width * (menuWidth / 100);

    return (
      <View style={[styles.appContainer, { width: screen.width }]}>
        {burgerIcon && (
          <BurgerIcon
            menuExpanded={menuExpanded}
            leftAligned={leftAligned}
            animationDuration={animationDuration}
            onPress={this.props.onPress}
            burgerIconColor1={burgerIconColor1}
            burgerIconColor2={burgerIconColor2}
            burgerWidth={burgerWidth}
            burgerIconStyles={burgerIconStyles}
          />
        )}
        <Animated.View
          style={[
            drawerAnimation(),
            styles.drawer,
            {
              width: DRAWER_WIDTH,
              opacity: fadeAnim
            }
          ]}
        >
          {menuComponent}
        </Animated.View>
        <View style={styles.container}>
          <View style={styles.overlay}>
            {menuExpanded && overlay && (
              <Animated.View
                style={[styles.overlay, { opacity: overlayAnim }]}
              />
            )}
            {children}
          </View>
          {children}
        </View>
      </View>
    );
  };

  toggleMenu = () => {
    this.setState({ menuExpanded: !this.state.menuExpanded });
  };

  renderPush = () => {
    const {
      children,
      menuComponent,
      menuWidth,
      menuExpanded,
      overlay,
      overlayOpacity,
      animationDuration,
      burgerIcon,
      burgerIconColor1,
      burgerIconColor2,
      leftAligned,
      burgerIconStyles,
      burgerWidth
    } = this.props;
    const { rightOffsetPush, leftOffset } = this.state;
    const DRAWER_WIDTH = screen.width * (menuWidth / 100);

    const containerAnimation = () => {
      if (leftAligned) {
        return {
          transform: [{ translateX: leftOffset }]
        };
      } else {
        return {
          transform: [{ translateX: rightOffsetPush }]
        };
      }
    };

    const drawerPosition = () => {
      if (leftAligned) {
        return {
          left: -DRAWER_WIDTH
        };
      } else {
        return {
          left: screen.width
        };
      }
    };

    return (
      <View style={styles.appContainer}>
        {burgerIcon && (
          <BurgerIcon
            menuExpanded={menuExpanded}
            leftAligned={leftAligned}
            animationDuration={animationDuration}
            onPress={this.props.onPress}
            burgerIconColor1={burgerIconColor1}
            burgerIconColor2={burgerIconColor2}
            burgerWidth={burgerWidth}
            burgerIconStyles={burgerIconStyles}
          />
        )}
        <Animated.View
          style={[
            containerAnimation(),
            styles.appContainer,
            {
              width: screen.width + DRAWER_WIDTH
            }
          ]}
        >
          <View
            style={[
              drawerPosition(),
              styles.drawer,
              {
                width: DRAWER_WIDTH
              }
            ]}
          >
            {menuComponent}
          </View>
          <View style={styles.container}>
            <View style={styles.overlay}>
              {menuExpanded && overlay && (
                <View style={{ ...styles.overlay, opacity: overlayOpacity }} />
              )}
              {children}
            </View>
            {children}
          </View>
        </Animated.View>
      </View>
    );
  };

  render() {
    const { push } = this.props;
    return push ? this.renderPush() : this.renderOverlay();
  }
}

SideMenu.defaultProps = {
  menuExpanded: false,
  menuWidth: 80,
  animationDuration: 300,
  fade: false,
  overlay: false,
  overlayOpacity: 0.4,
  leftAligned: false,
  push: false,
  burgerIcon: false,
  burgerIconColor1: "#ffffff",
  burgerWidth: 35
};

SideMenu.propTypes = {
  menuExpanded: PropTypes.bool,
  menuComponent: PropTypes.object,
  menuWidth: PropTypes.number,
  animationDuration: PropTypes.number,
  fade: PropTypes.bool,
  overlay: PropTypes.bool,
  overlayOpacity: PropTypes.number,
  leftAligned: PropTypes.bool,
  push: PropTypes.bool,
  burgerIcon: PropTypes.bool,
  burgerIconColor1: PropTypes.string,
  burgerIconColor2: PropTypes.string,
  burgerIconStyles: PropTypes.object,
  burgerWidth: PropTypes.number
};

const styles = StyleSheet.create({
  appContainer: {
    position: "absolute",
    left: 0,
    top: 0
  },
  container: {
    position: "absolute",
    left: 0,
    width: screen.width,
    height: screen.height,
    zIndex: 0
  },
  drawer: {
    position: "absolute",
    height: screen.height,
    zIndex: 1
  },
  overlay: {
    position: "absolute",
    left: 0,
    width: screen.width,
    height: screen.height,
    zIndex: 0,
    backgroundColor: "#000000",
    zIndex: 999
  }
});

export default SideMenu;
